pyccel/parser/semantic.py

```diff
@@ -870,6 +904,92 @@ def _handle_function(self, expr, func, args, **settings):
                             expr, func.is_elemental)
             return new_expr
 
+    def _handle_kernel(self, expr, func, args, **settings):
+        """
+        Create a FunctionCall or an instance of a PyccelInternalFunction
+        from the function information and arguments
+
+        Parameters
+        ==========
+        expr : PyccelAstNode
+               The expression where this call is found (used for error output)
+        func : FunctionDef instance, Interface instance or PyccelInternalFunction type
+               The function being called
+        args : tuple
+               The arguments passed to the function
+
+        Returns
+        =======
+        new_expr : FunctionCall or PyccelInternalFunction
+        """
+        if 'kernel' not in func.decorators:
+            errors.report(INVALID_FUNCTION_CALL,
+                        symbol = expr,
+                        severity = 'fatal')
+        if isinstance(func, PyccelFunctionDef):
+            func = func.cls_name
+            args, kwargs = split_positional_keyword_arguments(*args)
+            for a in args:
+                if getattr(a,'dtype',None) == 'tuple':
+                    self._infere_type(a, **settings)
+            for a in kwargs.values():
+                if getattr(a,'dtype',None) == 'tuple':
+                    self._infere_type(a, **settings)
+
+            try:
+                new_expr = func(*args, **kwargs)
+            except TypeError:
+                errors.report(UNRECOGNISED_FUNCTION_CALL,
+                        symbol = expr,
+                        severity = 'fatal')
+
+            return new_expr
+        else:
+            if isinstance(func, FunctionDef) and len(args) > len(func.arguments):
+                errors.report("Too many arguments passed in function call",
+                        symbol = expr,
+                        severity='fatal')
+            # TODO : type check the NUMBER OF BLOCKS 'numBlocks' and threads per block 'tpblock'
+            if not isinstance(expr.numBlocks, LiteralInteger):
+                # expr.numBlocks could be invalid type, or PyccelSymbol
+                if isinstance(expr.numBlocks, PyccelSymbol):
+                    numBlocks = self.get_variable(expr.numBlocks)
+                    if not isinstance(numBlocks.dtype, NativeInteger):
+                        errors.report(INVALID_KERNEL_CALL_BP_GRID,
+                        symbol = expr,
+                        severity='error')
+                else:
+                    errors.report(INVALID_KERNEL_CALL_BP_GRID,
+                        symbol = expr,
+                        severity='error')
```

Comment:
I find this quite hard to follow. I suggest:
```suggestion
                if isinstance(expr.numBlocks, PyccelSymbol):
                    numBlocks = self.get_variable(expr.numBlocks)
                if not isinstance(numBlocks.dtype, NativeInteger):
                    errors.report(INVALID_KERNEL_CALL_BP_GRID,
                        symbol = expr,
                        severity='error')
```

Reply:
Hmm, you're right. I didn't look closely enough and assumed it was mentioned at the start. Accessing `expr.numBlocks` (via a property) is more costly than accessing a variable directly so usually we prefer defining a variable once rather than using a property multiple times (I need to write this in a doc) . I missed that this wasn't the case here.
```suggestion
            numBlocks = expr.numBlocks
            if isinstance(numBlocks, PyccelSymbol):
                    numBlocks = self.get_variable(numBlocks)
            if not isinstance(numBlocks.dtype, NativeInteger):
                    errors.report(INVALID_KERNEL_CALL_BP_GRID,
                        symbol = expr,
                        severity='error')
```

Reply:
Doc written. See https://github.com/pyccel/pyccel/pull/1357